src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_aggregate.rs

```diff
@@ -0,0 +1,158 @@
+
+/// Heuristic optimizer runs in a bottom-up recursion fashion. If we match a plan like
+/// Filter-Aggregate-* and push down filter to Filter(Optional)-Aggregate-Filter-*, this will not
+/// work. RuleSplitAggregate will be applied first, since it's bottom up, then this rule, which
+/// cause the plan be like Filter(Optional)-Aggregate-Filter-Aggregate-*, which makes no sense.
+/// Hence we match 2 bundled Aggregate Ops:
+///
+/// Input:  Filter
+///           \
+///          Aggregate(Final)
+///            \
+///          Aggregate(Partial)
+///             \
+///              *
+///
+/// Output: Filter(Optional)
+///           \
+///          Aggregate(Final)
+///             \
+///            Aggregate(Partial)
+///               \
+///              Filter
+///                \
+///                 *
+pub struct RulePushDownFilterAggregate {
+    id: RuleID,
+    pattern: SExpr,
+}
+
+impl RulePushDownFilterAggregate {
+    pub fn new() -> Self {
+        Self {
+            id: RuleID::PushDownFilterAggregate,
+            pattern: SExpr::create_unary(
+                PatternPlan {
+                    plan_type: RelOp::Filter,
+                }
+                .into(),
+                SExpr::create_unary(
+                    PatternPlan {
+                        plan_type: OpAggregate,
+                    }
+                    .into(),
+                    SExpr::create_unary(
+                        PatternPlan {
+                            plan_type: OpAggregate,
+                        }
+                        .into(),
+                        SExpr::create_leaf(PatternPlan { plan_type: Pattern }.into()),
+                    ),
+                ),
+            ),
+        }
+    }
+}
+
+impl Rule for RulePushDownFilterAggregate {
+    fn id(&self) -> RuleID {
+        self.id
+    }
+
+    fn apply(&self, s_expr: &SExpr, state: &mut TransformResult) -> common_exception::Result<()> {
+        let filter: Filter = s_expr.plan().clone().try_into()?;
+        if filter.is_having {
+            let agg_parent = s_expr.child(0)?;
+            let agg_parent_plan: Aggregate = agg_parent.plan().clone().try_into()?;
+            let agg_child = agg_parent.child(0)?;
+            let agg_child_plan: Aggregate = agg_child.plan().clone().try_into()?;
+            if agg_parent_plan.mode == AggregateMode::Final
+                && agg_child_plan.mode == AggregateMode::Partial
+            {
+                let mut push_predicates = vec![];
+                let mut remaining_predicates = vec![];
+                for predicate in filter.predicates {
+                    let used_columns = predicate.used_columns();
+                    let mut pushable = true;
+                    for col in used_columns {
+                        if !agg_parent_plan.group_columns()?.contains(&col) {
+                            pushable = false;
+                            break;
+                        }
+                    }
```

Comment:
This is not correct if the aggregate function is referenced in `HAVING` clause, for example:
```sql
select a from t group by a having sum(b) > 1;
```

I suggest to follow this approach:
https://github.com/datafuselabs/databend/blob/a11deba5b493f9138ba2266835dca4ea8f590c59/src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs#L77-L92

Reply:
Thanks for taking a look! 
For `select a from t group by a having sum(b) > 1;`
I think in this case nothing will be pushed down,
```
mysql> explain select a from t1 group by a having sum(b) > 1;
+--------------------------------------------------------------+
| explain                                                      |
+--------------------------------------------------------------+
| Filter                                                       |
| ├── filters: [is_true(gt(sum(b) (#2), 1))]                   |
| ├── estimated rows: 0.00                                     |
| └── AggregateFinal                                           |
|     ├── group by: [a]                                        |
|     ├── aggregate functions: [sum(b)]                        |
|     ├── estimated rows: 0.00                                 |
|     └── AggregatePartial                                     |
|         ├── group by: [a]                                    |
|         ├── aggregate functions: [sum(b)]                    |
|         ├── estimated rows: 0.00                             |
|         └── TableScan                                        |
|             ├── table: default.default.t1                    |
|             ├── read rows: 0                                 |
|             ├── read bytes: 0                                |
|             ├── partitions total: 0                          |
|             ├── partitions scanned: 0                        |
|             ├── push downs: [filters: [], limit: NONE]       |
|             └── estimated rows: 0.00                         |
+--------------------------------------------------------------+
```
Another example:
```
mysql> explain select a from t1 group by a having sum(b) > 1 and a > 1;
+----------------------------------------------------------------------------------------+
| explain                                                                                |
+----------------------------------------------------------------------------------------+
| Filter                                                                                 |
| ├── filters: [is_true(gt(sum(b) (#2), 1))]                                             |
| ├── estimated rows: 0.00                                                               |
| └── AggregateFinal                                                                     |
|     ├── group by: [a]                                                                  |
|     ├── aggregate functions: [sum(b)]                                                  |
|     ├── estimated rows: 0.00                                                           |
|     └── AggregatePartial                                                               |
|         ├── group by: [a]                                                              |
|         ├── aggregate functions: [sum(b)]                                              |
|         ├── estimated rows: 0.00                                                       |
|         └── Filter                                                                     |
|             ├── filters: [gt(group_item (#0), 1)]                                      |
|             ├── estimated rows: 0.00                                                   |
|             └── TableScan                                                              |
|                 ├── table: default.default.t1                                          |
|                 ├── read rows: 0                                                       |
|                 ├── read bytes: 0                                                      |
|                 ├── partitions total: 0                                                |
|                 ├── partitions scanned: 0                                              |
|                 ├── push downs: [filters: [gt(group_item (#0), 1)], limit: NONE]       |
|                 └── estimated rows: 0.00                                               |
+----------------------------------------------------------------------------------------+
```
Can you please explain a bit more on what's the problem?

Reply:
Oh, I misunderstood your code. It's fine. Never mind.